### PR TITLE
Initial Commit for DLG Crash Info

### PR DESCRIPTION
**Pull request by jonkurtzjk**
Please review this DCI module for general release.

### CODE REVIEW

File: features/dlg_crash_info/src/dci/dci_config.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
+#ifndef _DCI_CONFIG_H_
+#define _DCI_CONFIG_H_
+
+
+/*
+***********************WARNING!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!*****************************************
+* This should be undefined once used in production.  The purpose of this is to test the module with RAM 
+* use ONLY.   This modifies the reset procedure to do a soft reset back into RAM and could potentially
+* cause isses iwth the scatterload.
```
Comment (pkmurphy-renesas) [inline features/dlg_crash_info/src/dci/dci_config.h:9]: Typo: iwth -> with

File: features/dlg_crash_info/README.md [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+This example uses the UUID from the DCI Service and appends a uint8_t for the number of resets.
+
+***Advertising Data***
+
+| AD # | AD Length | AD Type                          | AD Data                                               |
+|------|-----------|----------------------------------|-------------------------------------------------------|
+|  1   |     2     | GAP_AD_TYPE_FLAGS                | 0x06 (BR/EDR Not Supported, LE Gerneral Discoverable) |
+|  2   |     18    | GAP_AD_TYPE_SERVICE_128_BIT_DATA | <DCI Sevice UUID, Number of Resets>                   |
+
+![advertising data](assets/advertising_data.PNG)
```
Comment (pkmurphy-renesas) [inline features/dlg_crash_info/README.md:193]: Github not displaying image as README uses uppercase extension (.PNG) but file uses a lowercase extension (.png)

File: features/dlg_crash_info/README.md [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+***Scan Response Data***
+| AD # | AD Length | AD Type                          | AD Data |
+|------|-----------|----------------------------------|---------|
+|  1   |     4     | GAP_AD_TYPE_COMPLETE_NAME        | "DCI"   |
+
+![scan response data](assets/scan_rsp_data.PNG)
```
Comment (pkmurphy-renesas) [inline features/dlg_crash_info/README.md:200]: Github not displaying image as README uses uppercase extension (.PNG) but file uses a lowercase extension (.png)

File: features/dlg_crash_info/README.md [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+Each data type used in this application, can be referred to in dci_types.h.
+dci_get_data returns a type of dci_info_t
+
+```c
+typedef __PACKED_STRUCT
+{
+	uint8_t			header;
+	uint16_t 		length;
+	dci_data_t	    *data;
```
Comment (pkmurphy-renesas) [inline features/dlg_crash_info/README.md:104]: Field name alignment is off when viewing README on Github

File: features/dlg_crash_info/README.md [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+- data:  dci_data_t data type
+
+```c
+typedef __PACKED_STRUCT
+{
+	dci_reset_reason_t		    last_reset_reason;
+	uint8_t						num_resets;
+	dci_fault_info_t			fault_data[DCI_NUM_RESETS_STORED];
+	
+}dci_data_t;
```
Comment (pkmurphy-renesas) [inline features/dlg_crash_info/README.md:120]: Field name alignment is off when viewing README on Github

File: features/dlg_crash_info/README.md [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+> __Note:__ 
+LF_PLATFORM_RESET is currently not implemented.  This typically is a result of memory allocation issues on the heap, and is escalated into an NMI via an ASSERT_ERROR(0).  This should be evident in the call_trace when parsed.
+
+```c
+typedef __PACKED_STRUCT
+{
+	uint8_t 					data_valid;
+	uint32_t					epoch;
+	dci_last_fault_hdlr_t		fault_handler;
+	cortex_m0_stack_frame_t		stack_frame;
```
Comment (pkmurphy-renesas) [inline features/dlg_crash_info/README.md:160]: Field name alignment is off when viewing README on Github

File: features/dlg_crash_info/README.md [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+    > __Note:__ 
+    Refer to **Section 10** in [Get Started tutorial](http://lpccs-docs.dialog-semiconductor.com/Tutorial_SDK6/serial_port.html) for more information on enabling the UART for debugging purposes.
+
+4.  On boot, you will see a printout of number of resets along with reset issues.  Press SW2 on the motherboard to induce an error.  The terminal will provide two printouts: 1) The original boot message with no errors 2) The crash information log after the error is created.
+
+    ![Terminal_Error](assets/terminal_capture.PNG)
+
+
+5.  In addition to the UART printout, advertising data, along with a custom service is enabled to retrive the data.  A seperate tool has been created for the DA14531, as a central, to retrieve this data using GTL and python. The BLE interface is described below, and the ReadMe.md can for central_crash_info can be referrred to for operation.
```
Comment (pkmurphy-renesas) [inline features/dlg_crash_info/README.md:69]: Suggest linking to the specific example, or at least to the tool repo (the py_ble_manager repo will will be public soon):
https://github.com/dialog-semiconductor/py_ble_manager/tree/main/examples/central/3_central_crash_info

Comment (pkmurphy-renesas) [inline features/dlg_crash_info/README.md:69]: Remove the first 'can' from this sentence: "the ReadMe.md **can** for central_crash_info can be referrred to for operation"

Comment (pkmurphy-renesas) [inline features/dlg_crash_info/README.md:69]: Typo: referrred -> referred

**Comment by pkmurphy-renesas** [comment]
Had no issues building/running for compiler 6. Could not build for compiler 5 (v5.06 update 7 (build 960)).

**Comment by blauret** [comment]
It would be nice to add the cmake build to the example.

**Comment by massimilian-89** [comment]
Would you change the figures format and style in readme to have the same style for all the examples? you can find the SVG file in the other examples.

File: features/dlg_crash_info/README.md (hunk 2 above)
Comment (jonkurtzjk) [reply features/dlg_crash_info/README.md]: Fixed extension

File: features/dlg_crash_info/README.md (hunk 3 above)
Comment (jonkurtzjk) [reply features/dlg_crash_info/README.md]: Fixed png extension

File: features/dlg_crash_info/README.md (hunk 4 above)
Comment (jonkurtzjk) [reply features/dlg_crash_info/README.md]: Fixed Alignment

File: features/dlg_crash_info/README.md (hunk 5 above)
Comment (jonkurtzjk) [reply features/dlg_crash_info/README.md:118]: fixed alignment

File: features/dlg_crash_info/README.md (hunk 7 above)
Comment (jonkurtzjk) [reply features/dlg_crash_info/README.md]: Fixed grammar

Comment (jonkurtzjk) [reply features/dlg_crash_info/README.md]: Fixed Spelling

Comment (jonkurtzjk) [reply features/dlg_crash_info/README.md]: I added a link in the ReadMe.md in the opening paragraph describing the BLE usage.

**Comment by jonkurtzjk** [comment]
> Would you change the figures format and style in readme to have the same style for all the examples? you can find the SVG file in the other examples.

Hi, this has never been communicated publicly that we were switching to a standard for the ReadMe.md.   Which examples and can you point me to the internal documentation that describes this process.

**Comment by massimilian-89** [comment]
There are some issues I need to mention. Is there any particular reason to include both .h and .c file in the project?
In the user app, It is necessary to have the copyright license from Renesas on the files.

**Comment by massimilian-89** [comment]
in building process , I have this fatal error , did I do something wrong?

compiling user_crash_info.c...
linking...
Program Size: Code=16104 RO-data=2248 RW-data=80 ZI-data=7116  
FromELF: creating hex file...
After Build - User command #1: fromelf --bincombined --output=".\out_DA14531\Objects\dlg_crash_info_531.bin"  ".\out_DA14531\Objects\dlg_crash_info_531.axf"
After Build - User command #2: fromelf C:\Users\a5140624\Desktop\Initial\DA145xx_SDK\6.0.18.1182.1\projects\BLE_SDK6_examples\features\dlg_crash_info\Keil_5\out_DA14531\Objects\dlg_crash_info_531.axf --disassemble --interleave=source  --text -c --output=outfile.lst
ARM FromELF: Execution interrupted due to an illegal storage access.
Finished: 1 information, 0 warning, 0 error and 1 fatal error messages.
".\out_DA14531\Objects\dlg_crash_info_531.axf" - 0 Error(s), 0 Warning(s).
Build Time Elapsed:  00:00:18

**Comment by massimilian-89** [comment]
by this example we are supposed to see a log each time application crashes on terminal (in this case each time pushing reset button)?
If this is the case I cannot see it. Only on the first run I see such log ,not updating by each push.

**Comment by jonkurtzjk** [comment]
> There are some issues I need to mention. Is there any particular reason to include both .h and .c file in the project? In the user app, It is necessary to have the copyright license from Renesas on the files.

Yes, the APIs are then exposed and better for reference.  It is a personal peference, and more of the industry standard than not including them.  Having to 'toggle' the header file is 'clunky'

**Comment by jonkurtzjk** [comment]
> in building process , I have this fatal error , did I do something wrong?
> 
> compiling user_crash_info.c... linking... Program Size: Code=16104 RO-data=2248 RW-data=80 ZI-data=7116 FromELF: creating hex file... After Build - User command #1: fromelf --bincombined --output=".\out_DA14531\Objects\dlg_crash_info_531.bin" ".\out_DA14531\Objects\dlg_crash_info_531.axf" After Build - User command #2: fromelf C:\Users\a5140624\Desktop\Initial\DA145xx_SDK\6.0.18.1182.1\projects\BLE_SDK6_examples\features\dlg_crash_info\Keil_5\out_DA14531\Objects\dlg_crash_info_531.axf --disassemble --interleave=source --text -c --output=outfile.lst ARM FromELF: Execution interrupted due to an illegal storage access. Finished: 1 information, 0 warning, 0 error and 1 fatal error messages. ".\out_DA14531\Objects\dlg_crash_info_531.axf" - 0 Error(s), 0 Warning(s). Build Time Elapsed: 00:00:18

Please do a pull.  This is is an old error message.  The latest Keil does not like to create a listing file properly.  You can check the user Run 2 - on some computers I don't have errors, on others I do.  I still creates the assembly file just throws an error.

**Comment by jonkurtzjk** [comment]
> by this example we are supposed to see a log each time application crashes on terminal (in this case each time pushing reset button)? If this is the case I cannot see it. Only on the first run I see such log ,not updating by each push.

Are you loading it to flash prior?

**Comment by jonkurtzjk** [comment]
> > There are some issues I need to mention. Is there any particular reason to include both .h and .c file in the project? In the user app, It is necessary to have the copyright license from Renesas on the files.
> 
> Yes, the APIs are then exposed and better for reference. It is a personal peference, and more of the industry standard than not including them. Having to 'toggle' the header file is 'clunky'

One other comment.  These files are are config files as well, intended for the user to modify easily if needed.  If I do not include them, it won't be obvious to the users that there is flexibility in the API.

**Review by massimilian-89 — APPROVED** [review]

**Pull request merged**